{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: psycopg2-binary in /home/mohammad/.local/lib/python3.8/site-packages (2.9.10)\n",
      "Requirement already satisfied: pandas in /home/mohammad/.local/lib/python3.8/site-packages (1.3.4)\n",
      "Requirement already satisfied: tqdm in /home/mohammad/.local/lib/python3.8/site-packages (4.67.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3/dist-packages (from pandas) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/mohammad/.local/lib/python3.8/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: numpy>=1.17.3; platform_machine != \"aarch64\" and platform_machine != \"arm64\" and python_version < \"3.10\" in /home/mohammad/.local/lib/python3.8/site-packages (from pandas) (1.21.4)\n",
      "Requirement already satisfied: six>=1.5 in /home/mohammad/.local/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas) (1.17.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install psycopg2-binary pandas tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from AlzheimerTargets import targets\n",
    "from database_connection_info import user, password, host, port"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cytochrome P450 19A1',\n",
       " 'Muscarinic acetylcholine receptor M2',\n",
       " 'Muscarinic acetylcholine receptors; M2 & M3',\n",
       " 'Dopamine D3 receptor',\n",
       " 'Dopamine D4 receptor',\n",
       " 'Neprilysin',\n",
       " 'Plasminogen',\n",
       " 'Muscarinic acetylcholine receptor M1',\n",
       " 'Angiotensin II receptor',\n",
       " 'Serotonin 1a (5-HT1a) receptor',\n",
       " 'Matrix metalloproteinase 3',\n",
       " 'Matrix metalloproteinase-1',\n",
       " 'Matrix metalloproteinase 9',\n",
       " 'Matrix metalloproteinase-2',\n",
       " 'ADAM17',\n",
       " 'Peroxisome proliferator-activated receptor gamma',\n",
       " 'Receptor protein-tyrosine kinase erbB-2',\n",
       " 'Tyrosine-protein kinase LCK',\n",
       " 'Angiotensin II type 2 (AT-2) receptor',\n",
       " 'Serum albumin',\n",
       " 'Adenosine A2a receptor',\n",
       " 'Intercellular adhesion molecule-1',\n",
       " 'Androgen Receptor',\n",
       " 'Estrogen receptor',\n",
       " 'Purinergic receptor P2Y2',\n",
       " 'Excitatory amino acid transporter 1',\n",
       " 'Excitatory amino acid transporter 2',\n",
       " 'Excitatory amino acid transporter 3',\n",
       " 'Caspase-1',\n",
       " 'Caspase-3',\n",
       " 'Caspase-6',\n",
       " 'Caspase-8',\n",
       " 'Renin',\n",
       " 'Sigma opioid receptor',\n",
       " 'Nitric oxide synthase, inducible',\n",
       " 'Nitric-oxide synthase, endothelial',\n",
       " 'Nitric-oxide synthase, brain',\n",
       " 'Nitric oxide synthase (inducible and endothelial)',\n",
       " 'Nitric oxide sythases; iNOS & nNOS',\n",
       " 'P-selectin',\n",
       " 'Complement C1r',\n",
       " 'Type-1 angiotensin II receptor',\n",
       " 'Serotonin 2c (5-HT2c) receptor',\n",
       " 'Serotonin 2a (5-HT2a) receptor',\n",
       " 'Monoamine oxidase A',\n",
       " 'Butyrylcholinesterase',\n",
       " 'Acetylcholinesterase',\n",
       " 'Lipoprotein lipase',\n",
       " 'Dopamine D1 receptor',\n",
       " 'Cathepsin D',\n",
       " 'Cathepsin G',\n",
       " 'Cyclooxygenase-2',\n",
       " 'Cyclooxygenase-1',\n",
       " 'Cyclooxygenase',\n",
       " 'Alpha-2a adrenergic receptor',\n",
       " 'Dopamine transporter',\n",
       " 'Serotonin 6 (5-HT6) receptor',\n",
       " 'Calpain 1',\n",
       " 'Glutamate receptor ionotropic, AMPA 1',\n",
       " 'Glutamate receptor ionotropic, AMPA 2',\n",
       " 'Glutamate receptor ionotropic, AMPA 3',\n",
       " 'Cannabinoid CB2 receptor',\n",
       " 'Cathepsin B',\n",
       " 'Beta-secretase 1',\n",
       " 'Adrenergic receptor beta',\n",
       " 'Dopamine receptors; D3 & D4',\n",
       " '11-beta-hydroxysteroid dehydrogenase 1',\n",
       " 'Monoamine oxidase B',\n",
       " 'Tissue-type plasminogen activator',\n",
       " 'Cytochrome P450 2D6',\n",
       " 'HMG-CoA reductase',\n",
       " 'TNF-beta',\n",
       " 'Beta-2 adrenergic receptor',\n",
       " 'Phospholipase A2 group IIA',\n",
       " 'Serine/threonine-protein kinase RAF',\n",
       " 'Gamma-secretase',\n",
       " 'Estrogen receptor alpha',\n",
       " 'Estrogen receptor beta',\n",
       " 'Melatonin receptor 1A',\n",
       " 'Melatonin receptor',\n",
       " 'Arachidonate 5-lipoxygenase',\n",
       " 'Muscarinic acetylcholine receptor M2 and M4',\n",
       " 'Muscarinic acetylcholine receptor M2 and M5',\n",
       " 'Beta-glucocerebrosidase',\n",
       " 'Serum paraoxonase/arylesterase 1',\n",
       " 'Angiotensin-converting enzyme',\n",
       " 'Dopamine receptor',\n",
       " 'GABA-B receptor',\n",
       " 'Muscarinic acetylcholine receptor',\n",
       " 'Cytochrome P450 17A1',\n",
       " 'FK506 binding protein 12',\n",
       " 'Cyclin-dependent kinase 1',\n",
       " 'Amyloid-beta A4 protein',\n",
       " 'Farnesyl diphosphate synthase',\n",
       " 'Corticotropin releasing factor receptor 1',\n",
       " 'Cyclin-dependent kinase 5/CDK5 activator 1',\n",
       " 'Glycogen synthase kinase-3',\n",
       " 'Neuronal acetylcholine receptor; alpha4/beta2',\n",
       " 'Neuronal acetylcholine receptor protein alpha-7 subunit',\n",
       " 'Gamma-secretase subunit PEN-2',\n",
       " 'Erythropoietin receptor',\n",
       " 'Purine nucleoside phosphorylase',\n",
       " 'C-C chemokine receptor type 5',\n",
       " 'C-C chemokine receptor type 3',\n",
       " 'Urokinase-type plasminogen activator',\n",
       " 'C5a anaphylatoxin chemotactic receptor',\n",
       " 'Macrophage migration inhibitory factor',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 1',\n",
       " 'Tyrosine-protein kinase FYN',\n",
       " 'Serotonin transporter',\n",
       " 'Acyl coenzyme A:cholesterol acyltransferase 1',\n",
       " 'Peroxisome proliferator-activated receptor alpha',\n",
       " 'Peroxisome proliferator-activated receptor delta',\n",
       " 'P-glycoprotein 1',\n",
       " 'Tyrosine-protein kinase ABL',\n",
       " 'Neuronal acetylcholine receptor; alpha3/beta4',\n",
       " 'Glycogen synthase kinase-3 beta',\n",
       " 'Angiotensin-converting enzyme',\n",
       " 'Voltage-gated L-type calcium channel alpha-1C subunit',\n",
       " 'Glycogen synthase kinase-3 alpha',\n",
       " 'Monoamine oxidase',\n",
       " 'Beta-secretase (BACE)',\n",
       " 'Plasminogen activator inhibitor-1',\n",
       " 'Matrix metalloproteinase (2 and 3)',\n",
       " 'Transthyretin',\n",
       " 'Monoamine transporters; serotonin & dopamine',\n",
       " 'Matrix metalloproteinase 12',\n",
       " 'Mitogen-activated protein kinase; ERK1/ERK2',\n",
       " 'Ribosomal protein S6 kinase (P70S6K)',\n",
       " 'Serine/threonine-protein kinase AKT',\n",
       " 'Dual-specificity tyrosine-phosphorylation regulated kinase 1A',\n",
       " 'MAP kinase ERK1',\n",
       " 'MAP kinase ERK2',\n",
       " 'Tyrosine-protein kinase ABL',\n",
       " 'PPAR alpha/gamma',\n",
       " 'Insulin-like growth factor binding protein 3',\n",
       " 'TNF-alpha',\n",
       " 'Receptor-type tyrosine-protein phosphatase alpha',\n",
       " 'Glutamate receptor ionotropic AMPA',\n",
       " 'Prolyl endopeptidase',\n",
       " 'Insulin-like growth factor binding protein 2',\n",
       " 'Insulin-like growth factor binding protein 6',\n",
       " 'Ceramide glucosyltransferase',\n",
       " 'Insulin-like growth factor I receptor',\n",
       " 'Insulin receptor',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase; MEK1/2',\n",
       " 'Glyceraldehyde-3-phosphate dehydrogenase liver',\n",
       " 'Adenosine receptors; A2a & A3',\n",
       " 'Caspase-9',\n",
       " 'Vitamin D receptor',\n",
       " 'Vesicular acetylcholine transporter',\n",
       " 'Nerve growth factor receptor Trk-A',\n",
       " 'Dual specificity mitogen-activated protein kinase kinase 2',\n",
       " 'Subtilisin/kexin type 9',\n",
       " 'GAR transformylase',\n",
       " 'Anandamide amidohydrolase',\n",
       " 'Quinone reductase 1',\n",
       " 'Muscarinic acetylcholine receptor M1 and M3',\n",
       " 'Thioredoxin reductase',\n",
       " 'Caspase-7',\n",
       " 'Phenylethanolamine N-methyltransferase',\n",
       " 'Serotransferrin',\n",
       " 'CaM kinase II',\n",
       " 'Vascular endothelial growth factor A',\n",
       " 'Glutamine synthetase',\n",
       " 'Arachidonate 15-lipoxygenase',\n",
       " 'Complement factor H',\n",
       " 'Tumor necrosis factor receptor R1',\n",
       " 'G-protein coupled receptor kinase 2',\n",
       " 'Arachidonate 12-lipoxygenase',\n",
       " 'Phosphodiesterase 4B',\n",
       " 'Beta-2 adrenergic receptor and beta-3 adrenergic receptor',\n",
       " 'RXR alpha/PPAR gamma',\n",
       " 'Serine/threonine-protein kinase RAF and Dual specificity mitogen-activated protein kinase kinase 1 (Raf/MEK)',\n",
       " 'Glutathione S-transferase Pi',\n",
       " 'Cyclin-dependent kinase 5',\n",
       " 'Bcr/Abl fusion protein',\n",
       " 'Caspase-2',\n",
       " 'Caspase-4',\n",
       " 'Interleukin-8 receptor B',\n",
       " 'Neuronal acetylcholine receptor; alpha2/beta4',\n",
       " 'Superoxide dismutase',\n",
       " 'Lysosomal acid lipase/cholesteryl ester hydrolase',\n",
       " 'Neuronal acetylcholine receptor subunit alpha-3',\n",
       " 'Phosphatidylcholine-retinol O-acyltransferase',\n",
       " 'Cannabinoid receptor',\n",
       " 'Nitric-oxide synthase (endothelial and brain)',\n",
       " 'Neuronal acetylcholine receptor; alpha3/beta2',\n",
       " 'Beta secretase 2',\n",
       " 'Glutamate (NMDA) receptor subunit zeta 1',\n",
       " 'Cystic fibrosis transmembrane conductance regulator',\n",
       " 'Inositol 1,4,5-trisphosphate receptor type 1',\n",
       " 'Glutamate [NMDA] receptor subunit epsilon 1',\n",
       " 'Glutamate [NMDA] receptor subunit epsilon 2',\n",
       " 'Poly [ADP-ribose] polymerase-1',\n",
       " 'DNA polymerase beta',\n",
       " 'Choline acetylase',\n",
       " 'Catechol O-methyltransferase',\n",
       " 'Cathepsin (B and K)',\n",
       " 'C-X-C chemokine receptor type 3',\n",
       " 'Glutamate NMDA receptor; GRIN1/GRIN2A',\n",
       " 'Glutamate NMDA receptor; GRIN1/GRIN2B',\n",
       " 'Serine/threonine-protein kinase WEE1',\n",
       " 'Aryl hydrocarbon receptor',\n",
       " 'Collagenase',\n",
       " 'Aldehyde dehydrogenase',\n",
       " 'Receptor protein-tyrosine kinase erbB-4',\n",
       " 'PI3-kinase p110-alpha/p85-alpha',\n",
       " 'Cytochrome P450 1A1',\n",
       " 'Proenkephalin B',\n",
       " '5-lipoxygenase/FLAP',\n",
       " 'LDL receptor',\n",
       " 'Interleukin-8',\n",
       " 'Myeloperoxidase',\n",
       " 'cAMP-dependent protein kinase alpha-catalytic subunit',\n",
       " 'Prion protein',\n",
       " 'Low affinity neurotrophin receptor p75NTR',\n",
       " 'Cerebroside-sulfatase',\n",
       " 'Quinone reductase 2',\n",
       " 'Suv3 helicase',\n",
       " 'Ribosomal protein S6 kinase 1',\n",
       " 'Serum amyloid P-component',\n",
       " 'HLA class I histocompatibility antigen A-3',\n",
       " 'Acidic fibroblast growth factor',\n",
       " 'HLA class II histocompatibility antigen, DRB1-15 beta chain',\n",
       " 'Peptidyl-prolyl cis-trans isomerase NIMA-interacting 1',\n",
       " 'Insulin-like growth factor II receptor',\n",
       " 'PPAR delta/gamma',\n",
       " 'Neuronal acetylcholine receptor subunit alpha-2/beta-2',\n",
       " 'Tumour suppressor p53/oncoprotein Mdm2',\n",
       " 'Prostaglandin E synthase 2',\n",
       " 'Casein kinase I delta',\n",
       " 'Fatty acid binding protein muscle',\n",
       " 'ADAMTS4',\n",
       " 'ADAM12',\n",
       " 'ADAM10',\n",
       " 'Breakpoint cluster region protein',\n",
       " 'S-100 protein beta chain',\n",
       " 'Serine/threonine-protein kinase EEF2K',\n",
       " 'LXR-beta',\n",
       " 'Coagulation factor XIII',\n",
       " 'Kinesin-like protein 1',\n",
       " 'Eukaryotic translation initation factor',\n",
       " 'Guanine nucleotide-binding protein G(o), alpha subunit 1',\n",
       " 'CaM kinase II alpha',\n",
       " 'Ephrin type-A receptor 5',\n",
       " 'CaM kinase II gamma',\n",
       " 'CaM kinase I alpha',\n",
       " 'Serine/threonine-protein kinase PAK 1',\n",
       " 'NAD-dependent deacetylase sirtuin 1',\n",
       " 'Toll-like receptor 4',\n",
       " 'Phosphodiesterase 8B',\n",
       " 'Synaptic vesicular amine transporter',\n",
       " 'Apoptosis regulator Bcl-W',\n",
       " 'Coagulation factor V',\n",
       " 'ADAMTS1',\n",
       " 'Neurotrophic tyrosine kinase receptor type 2',\n",
       " 'Luteinizing hormone/Choriogonadotropin receptor',\n",
       " 'Heme oxygenase 1',\n",
       " 'Insulin',\n",
       " 'Muscarinic acetylcholine receptor M1/M5 chimeric protein',\n",
       " 'Catenin beta-1',\n",
       " 'Phosphoenolpyruvate carboxykinase cytosolic',\n",
       " 'Interferon-induced, double-stranded RNA-activated protein kinase',\n",
       " 'Alpha-1-antichymotrypsin',\n",
       " 'Ubiquitin carboxyl-terminal hydrolase isozyme L1',\n",
       " 'Breast cancer type 1 susceptibility protein',\n",
       " 'Alpha-synuclein',\n",
       " 'Segment polarity protein dishevelled homolog DVL-1',\n",
       " 'Nucleoside diphosphate kinase 1',\n",
       " 'Nucleoside diphosphate kinase 2',\n",
       " 'Cyclin-dependent kinase 5 regulatory subunit 1',\n",
       " 'PI3-kinase p85-alpha subunit',\n",
       " 'Ornithine aminotransferase, mitochondrial',\n",
       " 'Ephrin type-A receptor 1',\n",
       " 'Death-associated protein kinase 1',\n",
       " 'Serine/threonine-protein kinase 11',\n",
       " 'Serine/threonine-protein kinase MARK1',\n",
       " 'Serine/threonine-protein kinase 24',\n",
       " 'Protein tyrosine kinase 2 beta',\n",
       " 'Apolipoprotein A-I',\n",
       " 'Thioredoxin reductase 2',\n",
       " 'NT-3 growth factor receptor',\n",
       " 'Receptor tyrosine-protein kinase erbB-3',\n",
       " 'Interleukin-2',\n",
       " 'Presenilin 1',\n",
       " 'Cellular tumor antigen p53',\n",
       " 'DNA-(apurinic or apyrimidinic site) lyase',\n",
       " 'Neuronal acetylcholine receptor protein alpha-4 subunit',\n",
       " 'Glutathione transferase omega 1',\n",
       " 'Huntingtin',\n",
       " 'Macrophage scavenger receptor types I and II',\n",
       " 'Cyclic AMP-responsive element-binding protein 1',\n",
       " 'Disks large homolog 4',\n",
       " 'Tumor necrosis factor ligand superfamily member 10',\n",
       " 'Tumor necrosis factor ligand superfamily member 6',\n",
       " 'Arylamine N-acetyltransferase 1',\n",
       " 'Arylamine N-acetyltransferase 2',\n",
       " 'Toll-like receptor 9',\n",
       " 'GABA-B receptor 1',\n",
       " 'Tryptophan 5-hydroxylase 1',\n",
       " 'Glucagon',\n",
       " 'Non-receptor tyrosine-protein kinase TNK1',\n",
       " 'Histone acetyltransferase KAT5',\n",
       " 'Endoplasmic reticulum aminopeptidase 1',\n",
       " 'Adenylate cyclase type II',\n",
       " 'Aquaporin-4',\n",
       " 'Peroxiredoxin-1',\n",
       " 'Acid ceramidase',\n",
       " 'Heat shock 70 kDa protein 1',\n",
       " 'Dynamin-2',\n",
       " 'Transcription factor Sp1',\n",
       " 'Creatine kinase B-type',\n",
       " 'Dickkopf-related protein 1',\n",
       " 'Kallikrein 8',\n",
       " 'Leucine-rich repeat serine/threonine-protein kinase 2',\n",
       " 'Integrin alpha M',\n",
       " 'Elongation factor 2',\n",
       " 'Tumor necrosis factor receptor superfamily member 5',\n",
       " 'Tumor necrosis factor receptor superfamily member 1B',\n",
       " 'Toll-like receptor (TLR7/TLR9)',\n",
       " 'MIF/CD74 (Macrophage migration inhibitory factor and HLA-DR antigens-associated invariant chain)',\n",
       " 'Eukaryotic translation initiation factor 2 subunit 1',\n",
       " 'PI3-kinase p110-delta/p85-alpha',\n",
       " 'C-C motif chemokine 5',\n",
       " 'Transcriptional regulator ERG',\n",
       " 'Prelamin-A/C',\n",
       " 'Microtubule-associated protein tau',\n",
       " 'Endoplasmic reticulum-associated amyloid beta-peptide-binding protein',\n",
       " 'Peroxisome proliferator-activated receptor gamma/Nuclear receptor coactivator 1',\n",
       " 'Peroxisome proliferator-activated receptor gamma/Nuclear receptor corepressor 2',\n",
       " 'Ubiquitin carboxyl-terminal hydrolase BAP1',\n",
       " 'Peroxisome proliferator-activated receptor gamma/Nuclear receptor coactivator 2',\n",
       " 'Insulin-degrading enzyme',\n",
       " 'Peroxisome proliferator-activated receptor gamma/Nuclear receptor coactivator 3',\n",
       " 'Relaxin-3 receptor 1',\n",
       " 'Neuronal acetylcholine receptor subunit alpha-6/beta-2',\n",
       " 'C-C motif chemokine 2',\n",
       " 'Neuronal acetylcholine receptor protein beta-2 subunit',\n",
       " 'Peptidyl-prolyl cis-trans isomerase D',\n",
       " 'Glutathione S-transferase A4',\n",
       " 'Glutathione transferase omega 2',\n",
       " 'Platelet glycoprotein 4',\n",
       " 'G1/S-specific cyclin D1',\n",
       " '78 kDa glucose-regulated protein',\n",
       " 'Interleukin-6',\n",
       " 'Somatostatin receptor 1',\n",
       " 'Somatostatin receptor 2',\n",
       " 'Somatostatin receptor 3',\n",
       " 'Somatostatin receptor 4',\n",
       " 'Somatostatin receptor 5',\n",
       " \"Bis(5'-adenosyl)-triphosphatase\",\n",
       " 'Protein polybromo-1',\n",
       " 'Transforming growth factor beta-1',\n",
       " 'Kallikrein 7',\n",
       " 'Regulator of G-protein signaling 4',\n",
       " 'Protein-arginine deiminase type-2',\n",
       " 'Interleukin-1 beta',\n",
       " 'Scavenger receptor class B member 1',\n",
       " 'Islet amyloid polypeptide',\n",
       " 'Appetite-regulating hormone',\n",
       " 'Protoporphyrinogen oxidase',\n",
       " 'Natural resistance-associated macrophage protein 2',\n",
       " 'Proteasome subunit beta type-9',\n",
       " 'DNA polymerase nu',\n",
       " 'NEDD8-activating enzyme E1 regulatory subunit',\n",
       " 'Neuronal acetylcholine receptors; alpha-3/beta-4, alpha-5, beta-2',\n",
       " 'Kallikrein 6',\n",
       " 'Phosphatidylinositol 3,4,5-trisphosphate 3-phosphatase and dual-specificity protein phosphatase PTEN',\n",
       " 'Natural resistance-associated macrophage protein 1',\n",
       " 'Heat shock-related 70 kDa protein 2',\n",
       " 'ATP synthase subunit alpha, mitochondrial',\n",
       " 'Guanine nucleotide-binding protein G(s), subunit alpha',\n",
       " 'Neurogenic locus notch homolog protein 1',\n",
       " 'Methionine synthase',\n",
       " 'Ubiquitin carboxyl-terminal hydrolase 21',\n",
       " 'Presenilin 2',\n",
       " 'Advanced glycosylation end product-specific receptor',\n",
       " 'Tumour suppressor protein p53/Mdm4',\n",
       " 'mTORC1',\n",
       " 'High mobility group protein B1',\n",
       " 'Alpha-ketoglutarate-dependent dioxygenase FTO',\n",
       " 'TAR DNA-binding protein 43',\n",
       " 'Muscarinic acetylcholine receptors; M1 & M2',\n",
       " 'Neuronal acetylcholine receptor; alpha4/beta4',\n",
       " 'Adenosine receptors; A1 & A2',\n",
       " 'Dopamine D2 receptor and serotonin 1a receptor',\n",
       " 'Cholinesterases; ACHE & BCHE',\n",
       " 'Serotonin 2 receptors; 5-HT2a & 5-HT2c',\n",
       " 'Serotonin receptor (2b and 2c)',\n",
       " 'Neuronal acetylcholine receptor; alpha3/alpha6/beta2/beta3']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No exact match found for target: Neuronal acetylcholine receptor subunit alpha-2/beta-2\n",
      "Fetching data for target CHEMBL1978 (Cytochrome P450 19A1)...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing CHEMBL1978: 5000it [00:01, 4486.88it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb Cell 4\u001b[0m in \u001b[0;36m<cell line: 113>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=112'>113</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39m__name__\u001b[39m \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m__main__\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=113'>114</a>\u001b[0m     target_names \u001b[39m=\u001b[39m targets\n\u001b[0;32m--> <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=114'>115</a>\u001b[0m     df \u001b[39m=\u001b[39m get_chembl_ids_and_fetch_data(target_names)\n",
      "\u001b[1;32m/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb Cell 4\u001b[0m in \u001b[0;36mget_chembl_ids_and_fetch_data\u001b[0;34m(target_names)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=74'>75</a>\u001b[0m cursor\u001b[39m.\u001b[39mitersize \u001b[39m=\u001b[39m \u001b[39m1000\u001b[39m  \u001b[39m# Batch size for server-side cursor\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=75'>76</a>\u001b[0m cursor\u001b[39m.\u001b[39mexecute(query, (target_id,))\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=77'>78</a>\u001b[0m \u001b[39mfor\u001b[39;00m row \u001b[39min\u001b[39;00m tqdm(cursor, desc\u001b[39m=\u001b[39m\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mProcessing \u001b[39m\u001b[39m{\u001b[39;00mtarget_id\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m):\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=78'>79</a>\u001b[0m     info \u001b[39m=\u001b[39m {\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=79'>80</a>\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mTarget Name\u001b[39m\u001b[39m\"\u001b[39m: target_name,\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=80'>81</a>\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mTarget ChEMBL ID\u001b[39m\u001b[39m\"\u001b[39m: target_id,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=98'>99</a>\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mQED Weighted\u001b[39m\u001b[39m\"\u001b[39m: row[\u001b[39m16\u001b[39m]\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=99'>100</a>\u001b[0m     }\n\u001b[1;32m    <a href='vscode-notebook-cell:/home/mohammad/Documents/GitHub/LLMs_for_Drug_Discovery/DataRetrievalFromChEMBL.ipynb#W3sZmlsZQ%3D%3D?line=100'>101</a>\u001b[0m     all_data\u001b[39m.\u001b[39mappend(info)\n",
      "File \u001b[0;32m~/.local/lib/python3.8/site-packages/tqdm/std.py:1181\u001b[0m, in \u001b[0;36mtqdm.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1178\u001b[0m time \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_time\n\u001b[1;32m   1180\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1181\u001b[0m     \u001b[39mfor\u001b[39;00m obj \u001b[39min\u001b[39;00m iterable:\n\u001b[1;32m   1182\u001b[0m         \u001b[39myield\u001b[39;00m obj\n\u001b[1;32m   1183\u001b[0m         \u001b[39m# Update and possibly print the progressbar.\u001b[39;00m\n\u001b[1;32m   1184\u001b[0m         \u001b[39m# Note: does not call self.update(1) for speed optimisation.\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.8/encodings/utf_8.py:15\u001b[0m, in \u001b[0;36mdecode\u001b[0;34m(input, errors)\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[39m### Codec APIs\u001b[39;00m\n\u001b[1;32m     13\u001b[0m encode \u001b[39m=\u001b[39m codecs\u001b[39m.\u001b[39mutf_8_encode\n\u001b[0;32m---> 15\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdecode\u001b[39m(\u001b[39minput\u001b[39m, errors\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mstrict\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m     16\u001b[0m     \u001b[39mreturn\u001b[39;00m codecs\u001b[39m.\u001b[39mutf_8_decode(\u001b[39minput\u001b[39m, errors, \u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m     18\u001b[0m \u001b[39mclass\u001b[39;00m \u001b[39mIncrementalEncoder\u001b[39;00m(codecs\u001b[39m.\u001b[39mIncrementalEncoder):\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import psycopg2\n",
    "import pandas as pd\n",
    "from tqdm import tqdm \n",
    "\n",
    "def get_chembl_ids_and_fetch_data(target_names):\n",
    "    # Connect to PostgreSQL\n",
    "    conn = psycopg2.connect(\n",
    "        dbname=\"chembl_35\",\n",
    "        user=user,  \n",
    "        password=password,  \n",
    "        host=host,\n",
    "        port=port\n",
    "    )\n",
    "\n",
    "    chembl_ids = {}\n",
    "\n",
    "    # Step 1: Get ChEMBL IDs for target names (using SQL)\n",
    "    with conn.cursor() as cursor:\n",
    "        for target_name in target_names:\n",
    "            cursor.execute(\"\"\"\n",
    "                SELECT chembl_id, target_type, organism\n",
    "                FROM target_dictionary\n",
    "                WHERE pref_name ILIKE %s\n",
    "            \"\"\", (target_name,))\n",
    "            \n",
    "            results = cursor.fetchall()\n",
    "            if results:\n",
    "                chembl_ids[target_name] = [{\n",
    "                    \"target_chembl_id\": row[0],\n",
    "                    \"target_type\": row[1],\n",
    "                    \"organism\": row[2]\n",
    "                } for row in results]\n",
    "            else:\n",
    "                print(f\"No exact match found for target: {target_name}\")\n",
    "\n",
    "    # Step 2: Fetch data using optimized SQL joins\n",
    "    all_data = []\n",
    "    for target_name, target_info_list in chembl_ids.items():\n",
    "        for target_info in target_info_list:\n",
    "            target_id = target_info[\"target_chembl_id\"]\n",
    "            print(f\"Fetching data for target {target_id} ({target_name})...\")\n",
    "            \n",
    "            try:\n",
    "                # Single query to get all related data using joins\n",
    "                query = \"\"\"\n",
    "                    SELECT \n",
    "                        a.standard_value,\n",
    "                        a.standard_type AS activity_type,\n",
    "                        m.chembl_id AS molecule_chembl_id,\n",
    "                        m.max_phase,\n",
    "                        m.molecule_type,\n",
    "                        cs.canonical_smiles,\n",
    "                        cs.standard_inchi,\n",
    "                        cs.standard_inchi_key,\n",
    "                        cp.full_molformula,\n",
    "                        cp.full_mwt,\n",
    "                        cp.alogp,\n",
    "                        cp.aromatic_rings,\n",
    "                        cp.hba,\n",
    "                        cp.hbd,\n",
    "                        cp.rtb,\n",
    "                        cp.psa,\n",
    "                        cp.qed_weighted\n",
    "                    FROM activities a\n",
    "                    JOIN assays ass ON a.assay_id = ass.assay_id\n",
    "                    JOIN target_dictionary td ON ass.tid = td.tid\n",
    "                    JOIN molecule_dictionary m ON a.molregno = m.molregno\n",
    "                    JOIN compound_structures cs ON m.molregno = cs.molregno\n",
    "                    JOIN compound_properties cp ON m.molregno = cp.molregno\n",
    "                    WHERE td.chembl_id = %s\n",
    "                    AND a.standard_value IS NOT NULL\n",
    "                \"\"\"\n",
    "                \n",
    "                with conn.cursor(name=\"fetch_activities\") as cursor:\n",
    "                    cursor.itersize = 1000  # Batch size for server-side cursor\n",
    "                    cursor.execute(query, (target_id,))\n",
    "                    \n",
    "                    for row in tqdm(cursor, desc=f\"Processing {target_id}\"):\n",
    "                        info = {\n",
    "                            \"Target Name\": target_name,\n",
    "                            \"Target ChEMBL ID\": target_id,\n",
    "                            \"Target Type\": target_info[\"target_type\"],\n",
    "                            \"Organism\": target_info[\"organism\"],\n",
    "                            \"Molecule ChEMBL ID\": row[2],\n",
    "                            \"Activity Type\": row[1],\n",
    "                            \"Activity Value\": row[0],\n",
    "                            \"Canonical SMILES\": row[5],\n",
    "                            \"Standard InChI\": row[6],\n",
    "                            \"Standard InChI Key\": row[7],\n",
    "                            \"Max Phase\": row[3],\n",
    "                            \"Molecular Formula\": row[8],\n",
    "                            \"Molecular Weight\": row[9],\n",
    "                            \"AlogP\": row[10],\n",
    "                            \"Aromatic Rings\": row[11],\n",
    "                            \"HBA\": row[12],\n",
    "                            \"HBD\": row[13],\n",
    "                            \"Rotatable Bonds\": row[14],\n",
    "                            \"Polar Surface Area (PSA)\": row[15],\n",
    "                            \"QED Weighted\": row[16]\n",
    "                        }\n",
    "                        all_data.append(info)\n",
    "\n",
    "            except Exception as e:\n",
    "                print(f\"Error retrieving data for target {target_id}: {e}\")\n",
    "                conn.rollback()\n",
    "            else:\n",
    "                conn.commit()\n",
    "\n",
    "    conn.close()\n",
    "    return pd.DataFrame(all_data)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    target_names = targets\n",
    "    df = get_chembl_ids_and_fetch_data(target_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target Name</th>\n",
       "      <th>Target ChEMBL ID</th>\n",
       "      <th>Target Type</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Molecule ChEMBL ID</th>\n",
       "      <th>Activity Type</th>\n",
       "      <th>Activity Value</th>\n",
       "      <th>Canonical SMILES</th>\n",
       "      <th>Standard InChI</th>\n",
       "      <th>Standard InChI Key</th>\n",
       "      <th>Max Phase</th>\n",
       "      <th>Molecular Formula</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>AlogP</th>\n",
       "      <th>Aromatic Rings</th>\n",
       "      <th>HBA</th>\n",
       "      <th>HBD</th>\n",
       "      <th>Rotatable Bonds</th>\n",
       "      <th>Polar Surface Area (PSA)</th>\n",
       "      <th>QED Weighted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cytochrome P450 19A1</td>\n",
       "      <td>CHEMBL1978</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL63297</td>\n",
       "      <td>IC50</td>\n",
       "      <td>190.0</td>\n",
       "      <td>Fc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1</td>\n",
       "      <td>InChI=1S/C17H14FN3O/c18-14-7-5-12(6-8-14)17(21...</td>\n",
       "      <td>RVFPOHIDBASNMF-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C17H14FN3O</td>\n",
       "      <td>295.32</td>\n",
       "      <td>3.01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>39.94</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cytochrome P450 19A1</td>\n",
       "      <td>CHEMBL1978</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL65328</td>\n",
       "      <td>IC50</td>\n",
       "      <td>590.0</td>\n",
       "      <td>Cc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1</td>\n",
       "      <td>InChI=1S/C18H17N3O/c1-13-6-8-14(9-7-13)18(21-1...</td>\n",
       "      <td>OZEGINYEQGNONL-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C18H17N3O</td>\n",
       "      <td>291.35</td>\n",
       "      <td>3.18</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>39.94</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cytochrome P450 19A1</td>\n",
       "      <td>CHEMBL1978</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL488</td>\n",
       "      <td>IC50</td>\n",
       "      <td>18500.0</td>\n",
       "      <td>CCC1(c2ccc(N)cc2)CCC(=O)NC1=O</td>\n",
       "      <td>InChI=1S/C13H16N2O2/c1-2-13(8-7-11(16)15-12(13...</td>\n",
       "      <td>ROBVIMPUHSLWNV-UHFFFAOYSA-N</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C13H16N2O2</td>\n",
       "      <td>232.28</td>\n",
       "      <td>1.35</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>72.19</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cytochrome P450 19A1</td>\n",
       "      <td>CHEMBL1978</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL291646</td>\n",
       "      <td>IC50</td>\n",
       "      <td>200.0</td>\n",
       "      <td>Clc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1</td>\n",
       "      <td>InChI=1S/C17H14ClN3O/c18-14-7-5-12(6-8-14)17(2...</td>\n",
       "      <td>UVOSSULRORMWHE-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C17H14ClN3O</td>\n",
       "      <td>311.77</td>\n",
       "      <td>3.52</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>39.94</td>\n",
       "      <td>0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cytochrome P450 19A1</td>\n",
       "      <td>CHEMBL1978</td>\n",
       "      <td>SINGLE PROTEIN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL304788</td>\n",
       "      <td>IC50</td>\n",
       "      <td>880.0</td>\n",
       "      <td>Clc1ccc(C(C2Cc3ccccc3O2)n2cncn2)c(Cl)c1</td>\n",
       "      <td>InChI=1S/C17H13Cl2N3O/c18-12-5-6-13(14(19)8-12...</td>\n",
       "      <td>ZSINOYZZEZDTLK-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C17H13Cl2N3O</td>\n",
       "      <td>346.22</td>\n",
       "      <td>4.18</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>39.94</td>\n",
       "      <td>0.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448774</th>\n",
       "      <td>Neuronal acetylcholine receptor; alpha3/alpha6...</td>\n",
       "      <td>CHEMBL2109233</td>\n",
       "      <td>PROTEIN COMPLEX</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL3329544</td>\n",
       "      <td>Emax</td>\n",
       "      <td>64.0</td>\n",
       "      <td>C[C@H]1C[C@@H]1C(=O)N1CC2CC(C1)N2</td>\n",
       "      <td>InChI=1S/C10H16N2O/c1-6-2-9(6)10(13)12-4-7-3-8...</td>\n",
       "      <td>INYNWEUTASDACG-KRTGUUSXSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C10H16N2O</td>\n",
       "      <td>180.25</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>32.34</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448775</th>\n",
       "      <td>Neuronal acetylcholine receptor; alpha3/alpha6...</td>\n",
       "      <td>CHEMBL2109233</td>\n",
       "      <td>PROTEIN COMPLEX</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL3329545</td>\n",
       "      <td>Emax</td>\n",
       "      <td>73.0</td>\n",
       "      <td>CC1(C)CC1C(=O)N1CC2CC(C1)N2</td>\n",
       "      <td>InChI=1S/C11H18N2O/c1-11(2)4-9(11)10(14)13-5-7...</td>\n",
       "      <td>LJEPPQATHWQMBV-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C11H18N2O</td>\n",
       "      <td>194.28</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>32.34</td>\n",
       "      <td>0.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448776</th>\n",
       "      <td>Neuronal acetylcholine receptor; alpha3/alpha6...</td>\n",
       "      <td>CHEMBL2109233</td>\n",
       "      <td>PROTEIN COMPLEX</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL3329546</td>\n",
       "      <td>Emax</td>\n",
       "      <td>110.0</td>\n",
       "      <td>O=C(C1CC1(F)F)N1CC2CC(C1)N2</td>\n",
       "      <td>InChI=1S/C9H12F2N2O/c10-9(11)2-7(9)8(14)13-3-5...</td>\n",
       "      <td>VXQLPXYMYHKASH-UHFFFAOYSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C9H12F2N2O</td>\n",
       "      <td>202.20</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>32.34</td>\n",
       "      <td>0.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448777</th>\n",
       "      <td>Neuronal acetylcholine receptor; alpha3/alpha6...</td>\n",
       "      <td>CHEMBL2109233</td>\n",
       "      <td>PROTEIN COMPLEX</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL4869892</td>\n",
       "      <td>IC50</td>\n",
       "      <td>135.0</td>\n",
       "      <td>CC[C@H](C)[C@@H]1NC(=O)[C@H](CCC(=O)O)NC(=O)[C...</td>\n",
       "      <td>InChI=1S/C68H103N23O21S4/c1-7-32(6)53-66(110)8...</td>\n",
       "      <td>ATDHPNPDBDEQAK-BCTYQQKGSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C68H103N23O21S4</td>\n",
       "      <td>1706.98</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448778</th>\n",
       "      <td>Neuronal acetylcholine receptor; alpha3/alpha6...</td>\n",
       "      <td>CHEMBL2109233</td>\n",
       "      <td>PROTEIN COMPLEX</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CHEMBL4872191</td>\n",
       "      <td>IC50</td>\n",
       "      <td>5.55</td>\n",
       "      <td>CC(C)C[C@@H]1NC(=O)[C@H](C)NC(=O)[C@H](CCCNC(=...</td>\n",
       "      <td>InChI=1S/C69H114N30O17S4/c1-32(2)19-41-58(108)...</td>\n",
       "      <td>XFBDKMSGGGCNAI-WVHVKVEVSA-N</td>\n",
       "      <td>None</td>\n",
       "      <td>C69H114N30O17S4</td>\n",
       "      <td>1764.13</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1448779 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Target Name Target ChEMBL ID  \\\n",
       "0                                     Cytochrome P450 19A1       CHEMBL1978   \n",
       "1                                     Cytochrome P450 19A1       CHEMBL1978   \n",
       "2                                     Cytochrome P450 19A1       CHEMBL1978   \n",
       "3                                     Cytochrome P450 19A1       CHEMBL1978   \n",
       "4                                     Cytochrome P450 19A1       CHEMBL1978   \n",
       "...                                                    ...              ...   \n",
       "1448774  Neuronal acetylcholine receptor; alpha3/alpha6...    CHEMBL2109233   \n",
       "1448775  Neuronal acetylcholine receptor; alpha3/alpha6...    CHEMBL2109233   \n",
       "1448776  Neuronal acetylcholine receptor; alpha3/alpha6...    CHEMBL2109233   \n",
       "1448777  Neuronal acetylcholine receptor; alpha3/alpha6...    CHEMBL2109233   \n",
       "1448778  Neuronal acetylcholine receptor; alpha3/alpha6...    CHEMBL2109233   \n",
       "\n",
       "             Target Type      Organism Molecule ChEMBL ID Activity Type  \\\n",
       "0         SINGLE PROTEIN  Homo sapiens        CHEMBL63297          IC50   \n",
       "1         SINGLE PROTEIN  Homo sapiens        CHEMBL65328          IC50   \n",
       "2         SINGLE PROTEIN  Homo sapiens          CHEMBL488          IC50   \n",
       "3         SINGLE PROTEIN  Homo sapiens       CHEMBL291646          IC50   \n",
       "4         SINGLE PROTEIN  Homo sapiens       CHEMBL304788          IC50   \n",
       "...                  ...           ...                ...           ...   \n",
       "1448774  PROTEIN COMPLEX  Homo sapiens      CHEMBL3329544          Emax   \n",
       "1448775  PROTEIN COMPLEX  Homo sapiens      CHEMBL3329545          Emax   \n",
       "1448776  PROTEIN COMPLEX  Homo sapiens      CHEMBL3329546          Emax   \n",
       "1448777  PROTEIN COMPLEX  Homo sapiens      CHEMBL4869892          IC50   \n",
       "1448778  PROTEIN COMPLEX  Homo sapiens      CHEMBL4872191          IC50   \n",
       "\n",
       "        Activity Value                                   Canonical SMILES  \\\n",
       "0                190.0                 Fc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1   \n",
       "1                590.0                 Cc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1   \n",
       "2              18500.0                      CCC1(c2ccc(N)cc2)CCC(=O)NC1=O   \n",
       "3                200.0                Clc1ccc(C(C2Cc3ccccc3O2)n2cncn2)cc1   \n",
       "4                880.0            Clc1ccc(C(C2Cc3ccccc3O2)n2cncn2)c(Cl)c1   \n",
       "...                ...                                                ...   \n",
       "1448774           64.0                  C[C@H]1C[C@@H]1C(=O)N1CC2CC(C1)N2   \n",
       "1448775           73.0                        CC1(C)CC1C(=O)N1CC2CC(C1)N2   \n",
       "1448776          110.0                        O=C(C1CC1(F)F)N1CC2CC(C1)N2   \n",
       "1448777          135.0  CC[C@H](C)[C@@H]1NC(=O)[C@H](CCC(=O)O)NC(=O)[C...   \n",
       "1448778           5.55  CC(C)C[C@@H]1NC(=O)[C@H](C)NC(=O)[C@H](CCCNC(=...   \n",
       "\n",
       "                                            Standard InChI  \\\n",
       "0        InChI=1S/C17H14FN3O/c18-14-7-5-12(6-8-14)17(21...   \n",
       "1        InChI=1S/C18H17N3O/c1-13-6-8-14(9-7-13)18(21-1...   \n",
       "2        InChI=1S/C13H16N2O2/c1-2-13(8-7-11(16)15-12(13...   \n",
       "3        InChI=1S/C17H14ClN3O/c18-14-7-5-12(6-8-14)17(2...   \n",
       "4        InChI=1S/C17H13Cl2N3O/c18-12-5-6-13(14(19)8-12...   \n",
       "...                                                    ...   \n",
       "1448774  InChI=1S/C10H16N2O/c1-6-2-9(6)10(13)12-4-7-3-8...   \n",
       "1448775  InChI=1S/C11H18N2O/c1-11(2)4-9(11)10(14)13-5-7...   \n",
       "1448776  InChI=1S/C9H12F2N2O/c10-9(11)2-7(9)8(14)13-3-5...   \n",
       "1448777  InChI=1S/C68H103N23O21S4/c1-7-32(6)53-66(110)8...   \n",
       "1448778  InChI=1S/C69H114N30O17S4/c1-32(2)19-41-58(108)...   \n",
       "\n",
       "                  Standard InChI Key Max Phase Molecular Formula  \\\n",
       "0        RVFPOHIDBASNMF-UHFFFAOYSA-N      None        C17H14FN3O   \n",
       "1        OZEGINYEQGNONL-UHFFFAOYSA-N      None         C18H17N3O   \n",
       "2        ROBVIMPUHSLWNV-UHFFFAOYSA-N       4.0        C13H16N2O2   \n",
       "3        UVOSSULRORMWHE-UHFFFAOYSA-N      None       C17H14ClN3O   \n",
       "4        ZSINOYZZEZDTLK-UHFFFAOYSA-N      None      C17H13Cl2N3O   \n",
       "...                              ...       ...               ...   \n",
       "1448774  INYNWEUTASDACG-KRTGUUSXSA-N      None         C10H16N2O   \n",
       "1448775  LJEPPQATHWQMBV-UHFFFAOYSA-N      None         C11H18N2O   \n",
       "1448776  VXQLPXYMYHKASH-UHFFFAOYSA-N      None        C9H12F2N2O   \n",
       "1448777  ATDHPNPDBDEQAK-BCTYQQKGSA-N      None   C68H103N23O21S4   \n",
       "1448778  XFBDKMSGGGCNAI-WVHVKVEVSA-N      None   C69H114N30O17S4   \n",
       "\n",
       "        Molecular Weight AlogP  Aromatic Rings  HBA  HBD  Rotatable Bonds  \\\n",
       "0                 295.32  3.01             3.0  4.0  0.0              3.0   \n",
       "1                 291.35  3.18             3.0  4.0  0.0              3.0   \n",
       "2                 232.28  1.35             1.0  3.0  2.0              2.0   \n",
       "3                 311.77  3.52             3.0  4.0  0.0              3.0   \n",
       "4                 346.22  4.18             3.0  4.0  0.0              3.0   \n",
       "...                  ...   ...             ...  ...  ...              ...   \n",
       "1448774           180.25  0.22             0.0  2.0  1.0              1.0   \n",
       "1448775           194.28  0.61             0.0  2.0  1.0              1.0   \n",
       "1448776           202.20  0.21             0.0  2.0  1.0              1.0   \n",
       "1448777          1706.98  None             NaN  NaN  NaN              NaN   \n",
       "1448778          1764.13  None             NaN  NaN  NaN              NaN   \n",
       "\n",
       "        Polar Surface Area (PSA) QED Weighted  \n",
       "0                          39.94         0.75  \n",
       "1                          39.94         0.74  \n",
       "2                          72.19         0.60  \n",
       "3                          39.94         0.74  \n",
       "4                          39.94         0.71  \n",
       "...                          ...          ...  \n",
       "1448774                    32.34         0.63  \n",
       "1448775                    32.34         0.66  \n",
       "1448776                    32.34         0.66  \n",
       "1448777                     None         None  \n",
       "1448778                     None         None  \n",
       "\n",
       "[1448779 rows x 20 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "782"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(set(list(df[\"Target ChEMBL ID\"]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'% Activity remaining',\n",
       " '% Control',\n",
       " '% Ctrl',\n",
       " '% Enzyme Activity (relative to DMSO controls, average n=2)',\n",
       " '% Inhibition of Control Agonist Response (Mean n=2)',\n",
       " '% Inhibition of Control Specific Binding',\n",
       " '% Inhibition of Control Specific Binding (Mean n=2)',\n",
       " '% Inhibition of Control Values',\n",
       " '% Inhibition of Control Values (Mean n=2)',\n",
       " '% Residual activity with Skepinone-L',\n",
       " '% binding',\n",
       " '% inhibition',\n",
       " '% maximum response',\n",
       " '% of Control Agonist Response (Mean n=2)',\n",
       " '% of activity',\n",
       " '% of control',\n",
       " '% of effect',\n",
       " '% of engagement',\n",
       " '% of inhibition',\n",
       " '% of inhibition or stimulation',\n",
       " '% of residual activity',\n",
       " '% residual kinase activity',\n",
       " '%Bound_Albumin',\n",
       " '%Inhib (Mean)',\n",
       " '%Max (Mean)',\n",
       " '%max',\n",
       " '-Delta G',\n",
       " '-Log ED50',\n",
       " '-Log K0.5',\n",
       " '-Log KB',\n",
       " '-Log KD',\n",
       " '-TdeltaS',\n",
       " '-log(RatioIC50)',\n",
       " '-logKa',\n",
       " '1/kmax',\n",
       " '1/kobsd-ka',\n",
       " 'A/Ad',\n",
       " 'A1 selectivity',\n",
       " 'A2',\n",
       " 'A2 selectivity',\n",
       " 'A50',\n",
       " 'AC50',\n",
       " 'ACAT activity',\n",
       " 'ACT',\n",
       " 'AE activity',\n",
       " 'AIDH activity',\n",
       " 'AQ',\n",
       " 'AT1 affinity',\n",
       " 'AUC',\n",
       " 'AbsAC40_uM',\n",
       " 'Absolute ratio',\n",
       " 'Accumulation',\n",
       " 'Acetylation',\n",
       " 'Activation',\n",
       " 'Activation (% of control)',\n",
       " 'Activity',\n",
       " 'Activity remaning',\n",
       " 'Activity_index',\n",
       " 'Affinity',\n",
       " 'Affinity constant',\n",
       " 'Affinity ratio',\n",
       " 'Aggregate',\n",
       " 'Agonism inflection point',\n",
       " 'Agonist',\n",
       " 'Agonist activity',\n",
       " 'Agonist effect',\n",
       " 'Agonist index',\n",
       " 'Agonistic activity',\n",
       " 'AlDH activity',\n",
       " 'Alloste ic enhan er (AE)',\n",
       " 'Alpha',\n",
       " 'Alpha max',\n",
       " 'Amount bound',\n",
       " 'Analgesic efficacy',\n",
       " 'Anogenital distance',\n",
       " 'Antagonism',\n",
       " 'Antagonist efficacy',\n",
       " 'Anti-uterotrophic effect',\n",
       " 'Antiovulatory activity',\n",
       " 'Antiuterotropic effect',\n",
       " 'Avoidance response',\n",
       " 'B-A/A-B',\n",
       " 'BEI',\n",
       " 'BMD',\n",
       " 'BP',\n",
       " 'BSI',\n",
       " 'Basal',\n",
       " 'Basal control',\n",
       " 'Behavioral effect',\n",
       " 'Beta2 duration',\n",
       " 'Binding',\n",
       " 'Binding activity',\n",
       " 'Binding affinity',\n",
       " 'Binding constant',\n",
       " 'Binding potency',\n",
       " 'Binding site fraction',\n",
       " 'Binding stoichiometry',\n",
       " 'Biological activity',\n",
       " 'Blockade',\n",
       " 'Blockage',\n",
       " 'Blood pressure',\n",
       " 'Bmax',\n",
       " 'Bmax1',\n",
       " 'Bmax2',\n",
       " 'Bone mineral density',\n",
       " 'C50',\n",
       " 'CC50',\n",
       " 'CCh',\n",
       " 'CD',\n",
       " 'CI',\n",
       " 'CI50',\n",
       " 'CL',\n",
       " 'COF',\n",
       " 'COI',\n",
       " 'CTF',\n",
       " 'Calculated Ratio',\n",
       " 'Carbonyl group',\n",
       " 'Cell number',\n",
       " 'Change',\n",
       " 'Change in body temperature',\n",
       " 'Competition factor',\n",
       " 'Compound recovery',\n",
       " 'Concentration',\n",
       " 'Control',\n",
       " 'Control compound IC50',\n",
       " 'Control of P-450 marker',\n",
       " 'Cooperativity',\n",
       " 'Covalent binding',\n",
       " 'DA release',\n",
       " 'DC50',\n",
       " 'DOSE',\n",
       " 'DR10',\n",
       " 'Decrease',\n",
       " 'Degradation',\n",
       " 'Delta G',\n",
       " 'Delta HDL',\n",
       " 'Delta MABP',\n",
       " 'Delta TC',\n",
       " 'Delta TM',\n",
       " 'Delta Tm',\n",
       " 'Delta blood pressure',\n",
       " 'Delta pKi',\n",
       " 'DeltaCp',\n",
       " 'Discrimination ratio',\n",
       " 'Displacement',\n",
       " 'Dissociation rate constant',\n",
       " 'Dopamine level',\n",
       " 'Dose',\n",
       " 'Drug degradation',\n",
       " 'Drug metabolism',\n",
       " 'Drug uptake',\n",
       " 'Duration',\n",
       " 'E',\n",
       " 'E2',\n",
       " 'E2 response',\n",
       " 'EC',\n",
       " 'EC0.25',\n",
       " 'EC1.5',\n",
       " 'EC12',\n",
       " 'EC15',\n",
       " 'EC150',\n",
       " 'EC200',\n",
       " 'EC24',\n",
       " 'EC25',\n",
       " 'EC2x',\n",
       " 'EC30',\n",
       " 'EC5',\n",
       " 'EC50',\n",
       " 'EC5x',\n",
       " 'EC70',\n",
       " 'EC80',\n",
       " 'EC90',\n",
       " 'ECmax',\n",
       " 'ED',\n",
       " 'ED delta',\n",
       " 'ED2',\n",
       " 'ED20',\n",
       " 'ED30',\n",
       " 'ED50',\n",
       " 'ED50 mg/animal/day',\n",
       " 'ED80',\n",
       " 'ED90',\n",
       " 'EDmin',\n",
       " 'EI',\n",
       " 'EPMR',\n",
       " 'ER',\n",
       " 'Effect',\n",
       " 'Efficacy',\n",
       " 'Efficacy ratio',\n",
       " 'Efficiency',\n",
       " 'Emax',\n",
       " 'Emax,diss',\n",
       " 'Emin',\n",
       " 'Enhancement',\n",
       " 'Enzyme activity',\n",
       " 'Enzyme inhibition',\n",
       " 'Epinephrine level',\n",
       " 'Estrogenic effect',\n",
       " 'Estrogenic potency',\n",
       " 'FC',\n",
       " 'FI',\n",
       " 'FI5',\n",
       " 'FICI',\n",
       " 'FSC recovery',\n",
       " 'Fe removed',\n",
       " 'Fibril formation',\n",
       " 'Flu intensity',\n",
       " 'Fluorescence intensity',\n",
       " 'Fm',\n",
       " 'Fold IC50',\n",
       " 'Fold activation',\n",
       " 'Fold change',\n",
       " 'Fold induction',\n",
       " 'Fold selectivity',\n",
       " 'Fold stimulation',\n",
       " 'Formation',\n",
       " 'Fu',\n",
       " 'Full inhibition',\n",
       " 'G-protein activation',\n",
       " 'GI',\n",
       " 'GI50',\n",
       " 'GTP-gammaS index',\n",
       " 'Gain',\n",
       " 'HSAI',\n",
       " 'HTS data',\n",
       " 'Hill coefficient',\n",
       " 'Hydrolysis',\n",
       " 'Hypotensive activity',\n",
       " 'I',\n",
       " 'I0.5',\n",
       " 'I10',\n",
       " 'IA',\n",
       " 'IA50',\n",
       " 'IAR',\n",
       " 'IAr',\n",
       " 'IC',\n",
       " 'IC100',\n",
       " 'IC20',\n",
       " 'IC25',\n",
       " 'IC33',\n",
       " 'IC5',\n",
       " 'IC50',\n",
       " 'IC50 ratio',\n",
       " 'IC60',\n",
       " 'IC80',\n",
       " 'IC90',\n",
       " 'IC95',\n",
       " 'ICmax',\n",
       " 'ID50',\n",
       " 'INH',\n",
       " 'IP',\n",
       " 'IS',\n",
       " 'ISA',\n",
       " 'IZ',\n",
       " 'Imax',\n",
       " 'Inactivation',\n",
       " 'Increase',\n",
       " 'Index',\n",
       " 'Induction',\n",
       " 'Infertility activity',\n",
       " 'Inhibition',\n",
       " 'Inhibition (% of control)',\n",
       " 'Inhibition (alpha = 0.01)',\n",
       " 'Inhibition (at concentration M)',\n",
       " 'Inhibition potency',\n",
       " 'Inhibition ratio',\n",
       " 'Inhibitory activity',\n",
       " 'Inhibitory activity ratio',\n",
       " 'Inhibitory ratio',\n",
       " 'Inihibition',\n",
       " 'Initial rates',\n",
       " 'Intrinsic activity',\n",
       " 'Intrinsic efficacy',\n",
       " 'Inverse agonism',\n",
       " 'K',\n",
       " 'K 0.5',\n",
       " 'K Bind',\n",
       " 'K H',\n",
       " 'K L',\n",
       " 'K SH',\n",
       " 'K aff',\n",
       " 'K app',\n",
       " 'K app ratio',\n",
       " 'K ass',\n",
       " 'K bind',\n",
       " 'K cat/Kinact',\n",
       " 'K diss',\n",
       " 'K inact',\n",
       " 'K inact/Ki',\n",
       " 'K obs',\n",
       " 'K obs / 1',\n",
       " 'K-1',\n",
       " 'K-2',\n",
       " 'K0.5',\n",
       " 'K1',\n",
       " 'K2',\n",
       " 'K6',\n",
       " 'KA',\n",
       " 'KB',\n",
       " 'KB app',\n",
       " 'KD app',\n",
       " \"KD'\",\n",
       " 'KD1',\n",
       " 'KD2',\n",
       " 'KH',\n",
       " 'KHOX',\n",
       " \"KI'\",\n",
       " 'KI50',\n",
       " 'KII',\n",
       " 'KIs',\n",
       " 'KL',\n",
       " 'KOX',\n",
       " 'KPCB',\n",
       " 'KR',\n",
       " 'KSV',\n",
       " 'KT4',\n",
       " 'Ka',\n",
       " 'Ka app',\n",
       " 'Ka(T4)',\n",
       " 'Kapp',\n",
       " 'Kb',\n",
       " 'Kbapp',\n",
       " 'Kc',\n",
       " 'Kcat',\n",
       " 'Kcat/Km',\n",
       " 'Kd',\n",
       " 'Kd,app',\n",
       " 'Kd1',\n",
       " 'Kd2',\n",
       " 'Kdapp',\n",
       " 'Kdiss',\n",
       " 'Ke',\n",
       " 'Ke(app)',\n",
       " 'Ki',\n",
       " 'Ki 0.5',\n",
       " 'Ki app',\n",
       " 'Ki ratio',\n",
       " \"Ki'\",\n",
       " \"Ki''\",\n",
       " 'Ki1',\n",
       " 'Ki2',\n",
       " 'KiA',\n",
       " 'KiAH',\n",
       " 'KiH',\n",
       " 'KiL',\n",
       " 'Kic',\n",
       " 'Kieq',\n",
       " 'Kii',\n",
       " 'Kina',\n",
       " 'Kinact',\n",
       " 'Kinact/Ki',\n",
       " 'Kis',\n",
       " 'Kiuc',\n",
       " 'Km',\n",
       " 'Km/Ki',\n",
       " 'Kobs',\n",
       " 'Koff',\n",
       " 'Kq',\n",
       " 'Kr',\n",
       " 'Kr/kmax',\n",
       " 'Ks',\n",
       " 'Ks app',\n",
       " 'Ks/i',\n",
       " 'LC50',\n",
       " 'LED',\n",
       " 'LS',\n",
       " 'Log 1/C',\n",
       " 'Log 1/KL',\n",
       " 'Log 1/Kd',\n",
       " 'Log 1/M',\n",
       " 'Log B/F',\n",
       " 'Log CDQR',\n",
       " 'Log I50',\n",
       " 'Log IC50',\n",
       " \"Log K'\",\n",
       " 'Log KH',\n",
       " 'Log KL',\n",
       " 'Log Kd',\n",
       " 'Log Ki',\n",
       " 'Log potency ratio',\n",
       " 'Loss',\n",
       " 'M2/M1 Index',\n",
       " 'M2/M1 index',\n",
       " 'MA',\n",
       " 'MAO deamination',\n",
       " 'MEC',\n",
       " 'MEC1.5',\n",
       " 'MEC2.0',\n",
       " 'MED',\n",
       " 'MET',\n",
       " 'MFI',\n",
       " 'MIC',\n",
       " 'MIC50',\n",
       " 'MIC90',\n",
       " 'MPO release',\n",
       " 'MRT',\n",
       " 'Max',\n",
       " 'Max PI',\n",
       " 'Max binding',\n",
       " 'Max cAMP',\n",
       " 'Max contraction',\n",
       " 'Max effect',\n",
       " 'Max inhibition',\n",
       " 'Max intensity',\n",
       " 'Max response',\n",
       " 'Max stimulation',\n",
       " 'Max uterine weight',\n",
       " 'Max. activation',\n",
       " 'Max. potentiation',\n",
       " 'Maximal activation',\n",
       " 'Maximal effect',\n",
       " 'Maximum response',\n",
       " 'Mean',\n",
       " 'Mean decrease',\n",
       " 'Mean duration',\n",
       " 'Mean fold stimulation',\n",
       " 'Mean weight',\n",
       " 'Mem Rec',\n",
       " 'Metabolism',\n",
       " 'Molar equivalence bound',\n",
       " 'Mole fraction',\n",
       " 'Monomer',\n",
       " 'Muscarinic agonist index',\n",
       " 'Muscarinic receptor affinity',\n",
       " 'NO formation',\n",
       " 'NOHA',\n",
       " 'NSB',\n",
       " 'No. of rats',\n",
       " 'No. of survivors',\n",
       " 'No. of uptake sites',\n",
       " 'NonRGS-EC50 (uM)',\n",
       " 'OD',\n",
       " 'Occ',\n",
       " 'Occupancy',\n",
       " 'Oligo shifted',\n",
       " 'Organ weight',\n",
       " 'Others',\n",
       " 'PB',\n",
       " 'PC50',\n",
       " 'PCMA antagonism',\n",
       " 'PD2',\n",
       " 'PF50',\n",
       " 'PI',\n",
       " 'PI metabolism',\n",
       " 'PPB',\n",
       " 'PR',\n",
       " 'Palpha',\n",
       " 'Papp',\n",
       " 'Partition ratio',\n",
       " 'Peak current',\n",
       " 'Pgp AbA/analogue ratio',\n",
       " 'Plasma selectivity',\n",
       " 'Potency',\n",
       " 'Potency ratio',\n",
       " 'Potency_Dyrk1a_uM',\n",
       " 'Potentiation inflection point',\n",
       " 'RA',\n",
       " 'RAC',\n",
       " 'RAC app',\n",
       " 'RBA',\n",
       " 'RBA data',\n",
       " 'RBA1',\n",
       " 'RBA2',\n",
       " 'RBI',\n",
       " 'RCI',\n",
       " 'RE',\n",
       " 'REP',\n",
       " 'RFU',\n",
       " 'RLU',\n",
       " 'RM',\n",
       " 'RP',\n",
       " 'RRT',\n",
       " 'RSA',\n",
       " 'RSA tR',\n",
       " 'RTP',\n",
       " 'RU',\n",
       " 'RUmax',\n",
       " 'Rate',\n",
       " 'Rate of hydrolysis',\n",
       " 'Rate of inhibition',\n",
       " 'Rate of oxidation',\n",
       " 'Rate of reduction',\n",
       " 'Ratio',\n",
       " 'Ratio EC50',\n",
       " 'Ratio ED50',\n",
       " 'Ratio IC50',\n",
       " 'Ratio Ki',\n",
       " 'Ratio Ki(app)',\n",
       " 'Ratio LC50/IC50',\n",
       " 'Ratio pIC50',\n",
       " 'Ratio pKi',\n",
       " 'RatioGI50',\n",
       " 'Ratio_ATPase activity',\n",
       " 'Ratio_Papp',\n",
       " 'Reactivation',\n",
       " 'Reactivity',\n",
       " 'Receptor occupied',\n",
       " 'Recovery',\n",
       " 'Reduction',\n",
       " 'Relative JC-1 accumulation',\n",
       " 'Relative Potency',\n",
       " 'Relative Selectivity',\n",
       " 'Relative activation',\n",
       " 'Relative activity',\n",
       " 'Relative activity of AChE',\n",
       " 'Relative affinity',\n",
       " 'Relative antiandrogenic activity',\n",
       " 'Relative binding affinity',\n",
       " 'Relative binding potency',\n",
       " 'Relative efficacy',\n",
       " 'Relative florescence',\n",
       " 'Relative intrinsic activity',\n",
       " 'Relative potency',\n",
       " 'Relative rate of hydrolysis',\n",
       " 'Relative response',\n",
       " 'Relative seminal vesicle weight',\n",
       " 'Relative uterine weight',\n",
       " 'Relative uterotrophic activity',\n",
       " 'Relative uterus weight',\n",
       " 'Release',\n",
       " 'Residual Activity',\n",
       " 'Residual activity',\n",
       " 'Residual peptide',\n",
       " 'Residual_activity',\n",
       " 'Response',\n",
       " 'Response unit',\n",
       " 'Retention_time',\n",
       " 'Reversal activity',\n",
       " 'Reversibility',\n",
       " 'Reversible',\n",
       " 'Ri',\n",
       " 'Rmax',\n",
       " 'S/N',\n",
       " 'S20,w',\n",
       " 'SB',\n",
       " 'SC50',\n",
       " 'SI',\n",
       " 'SS',\n",
       " 'Selectivity',\n",
       " 'Selectivity Index',\n",
       " 'Selectivity coefficient',\n",
       " 'Selectivity index',\n",
       " 'Selectivity ratio',\n",
       " 'Selectivity ratio S3/2',\n",
       " 'Serum lipoprotein cholesterol',\n",
       " 'Smax',\n",
       " 'Solubility',\n",
       " 'Specific activity',\n",
       " 'Specific binding',\n",
       " 'Specificity',\n",
       " 'Stability',\n",
       " 'Stimulation',\n",
       " 'Stimulations',\n",
       " 'Substrate specificity',\n",
       " 'Superoxide anion',\n",
       " 'Suppression',\n",
       " 'Survival',\n",
       " 'T',\n",
       " 'T/C',\n",
       " 'T1/2',\n",
       " 'T50',\n",
       " 'TDI',\n",
       " 'TIME',\n",
       " 'TN',\n",
       " 'TdeltaS',\n",
       " 'Tetramer',\n",
       " 'Thermal melting change',\n",
       " 'Time',\n",
       " 'Time to max. inhibition',\n",
       " 'Tlag',\n",
       " 'Tm',\n",
       " 'Tm shift',\n",
       " 'Total binding',\n",
       " 'Total binding inhibited',\n",
       " 'Total turnings',\n",
       " 'Tp control',\n",
       " 'Transactivation',\n",
       " 'Transcriptional activity',\n",
       " 'Transport',\n",
       " 'Turnover',\n",
       " 'Uptake ratio',\n",
       " 'Uptake/Binding ratio',\n",
       " 'Uterine activity',\n",
       " 'Uterine weight',\n",
       " 'Uterine wet weight',\n",
       " 'Uterotrophic effect',\n",
       " 'Uterotropic activity',\n",
       " 'Uterotropic effect',\n",
       " 'Uterus effect',\n",
       " 'V1/2',\n",
       " 'VDR affinity',\n",
       " 'Vaginal activity',\n",
       " 'Vasodilation',\n",
       " 'Vmax',\n",
       " 'Vmax / Km',\n",
       " 'Vmax(app)',\n",
       " 'Vmax/Km',\n",
       " 'Weight',\n",
       " 'Weight (mg/g)',\n",
       " 'Wet weight',\n",
       " 'XC50',\n",
       " 'Xi50',\n",
       " \"[EOP']infinity/[EOH]infinity\",\n",
       " 'alphaKi',\n",
       " 'cAMP',\n",
       " 'cAMP activity index',\n",
       " 'cAMP increase',\n",
       " 'cAMP production',\n",
       " 'cKi',\n",
       " 'cholesterol level',\n",
       " 'dTM',\n",
       " 'dTm',\n",
       " 'delta Tm',\n",
       " 'delta Tm Staurosporine positive control',\n",
       " 'deltaG',\n",
       " 'deltaH',\n",
       " 'deltaS',\n",
       " 'deltaT',\n",
       " 'deltaTm',\n",
       " 'effect',\n",
       " 'fEC50',\n",
       " 'fIC50',\n",
       " 'fIC90',\n",
       " 'fold selectivity',\n",
       " 'fpKi',\n",
       " 'inhibition (at 100uM)',\n",
       " 'k',\n",
       " 'k cat',\n",
       " 'k cat/Km',\n",
       " 'k eff',\n",
       " 'k eff (rel)',\n",
       " 'k obs / 1',\n",
       " \"k'/k'+1\",\n",
       " 'k-2',\n",
       " 'k/K',\n",
       " 'k2',\n",
       " \"k2'\",\n",
       " 'k2/Ki',\n",
       " 'k2c',\n",
       " 'k3',\n",
       " 'k3/Ki',\n",
       " 'k3/ki',\n",
       " 'kH',\n",
       " 'kHOX',\n",
       " 'kOX',\n",
       " 'kR',\n",
       " 'k_obs',\n",
       " 'k_off',\n",
       " 'k_on',\n",
       " 'ka',\n",
       " 'ka/Ki',\n",
       " 'ka/kr',\n",
       " 'kaKr/kr',\n",
       " 'kb',\n",
       " 'ki/Ki',\n",
       " 'kinact',\n",
       " 'kirr',\n",
       " 'kn',\n",
       " 'koff',\n",
       " 'kon',\n",
       " 'kr',\n",
       " 'log K',\n",
       " 'log Ks',\n",
       " 'log(%)',\n",
       " 'log(1/C)',\n",
       " 'log(10^6/IC50)',\n",
       " 'log(IA)',\n",
       " 'log(PB)',\n",
       " 'log(RBA)',\n",
       " 'log(Tau/KA)',\n",
       " 'log(activity)',\n",
       " 'log(ratio)',\n",
       " 'log2FC',\n",
       " 'logEC50',\n",
       " 'logEmax',\n",
       " 'logFu',\n",
       " 'logIC50',\n",
       " 'logKA',\n",
       " 'logKb',\n",
       " 'logKd',\n",
       " 'logKi',\n",
       " 'logRA',\n",
       " 'max activation',\n",
       " 'mortality',\n",
       " 'n',\n",
       " 'nH',\n",
       " 'nNOS activity',\n",
       " 'p alpha',\n",
       " 'p alphaKa',\n",
       " 'pA10',\n",
       " 'pA2',\n",
       " 'pA2 - pA10',\n",
       " 'pA2(app)',\n",
       " 'pA50',\n",
       " 'pAh',\n",
       " 'pAlpha',\n",
       " 'pC1',\n",
       " 'pC2A',\n",
       " 'pC30',\n",
       " 'pD2',\n",
       " 'pEC0.5',\n",
       " 'pEC0.5,diss',\n",
       " 'pEC5',\n",
       " 'pEC50',\n",
       " 'pEC50 diss',\n",
       " 'pI50',\n",
       " 'pIC25',\n",
       " 'pIC50',\n",
       " 'pK',\n",
       " 'pK1',\n",
       " 'pKA',\n",
       " 'pKB',\n",
       " 'pKD',\n",
       " 'pKa',\n",
       " 'pKact',\n",
       " 'pKb',\n",
       " 'pKi',\n",
       " 'pKi(uM)',\n",
       " 'pKi1-pKi2',\n",
       " 'pKic',\n",
       " 'pKinact',\n",
       " 'pKinc',\n",
       " 'pKm',\n",
       " 'pKp',\n",
       " 'pKx',\n",
       " 'pRA',\n",
       " 'pRA1',\n",
       " 'pRatio',\n",
       " 'pXC50',\n",
       " 'p[A50]',\n",
       " 'p[A]50',\n",
       " 'permeability',\n",
       " 'sigmaH',\n",
       " 'sigmaL',\n",
       " 'sigmaSH'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(list(df[\"Activity Type\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ce7cece8b74f42f7836e5e9b1330e461",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer_config.json:   0%|          | 0.00/315 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "842ac43880194ee8afcdfabb43fbe6e6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "vocab.txt:   0%|          | 0.00/306k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0c484ee7c2a24644bd6887e88bab8dfa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "tokenizer.json:   0%|          | 0.00/777k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b82ef293d0df4a90926f117f36a92357",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cb818d72b56c46ff97435a302af9c775",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "config.json:   0%|          | 0.00/638 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "12a9f551839546679ead17cc93535fa3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "model.safetensors:   0%|          | 0.00/439M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from transformers import BertTokenizerFast, BertModel\n",
    "checkpoint = 'unikei/bert-base-smiles'\n",
    "tokenizer = BertTokenizerFast.from_pretrained(checkpoint)\n",
    "model = BertModel.from_pretrained(checkpoint)\n",
    "\n",
    "example = 'O=C([C@@H](c1ccc(cc1)O)N)N[C@@H]1C(=O)N2[C@@H]1SC([C@@H]2C(=O)O)(C)C'\n",
    "tokens = tokenizer(example, return_tensors='pt')\n",
    "predictions = model(**tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[-1.3821,  0.0593, -0.1914,  ..., -2.0613, -0.4603, -1.1894],\n",
       "         [-0.9418, -0.1656,  0.2671,  ..., -0.9770, -0.0633, -1.9062],\n",
       "         [ 1.1226, -0.4002,  0.3521,  ...,  0.1618, -0.1386,  1.8086],\n",
       "         ...,\n",
       "         [ 0.6186,  0.3017,  0.1241,  ..., -0.1342,  0.0974, -0.0602],\n",
       "         [ 0.4130, -1.1119,  0.2806,  ...,  0.0944, -0.5532, -0.3501],\n",
       "         [ 0.5375,  0.7749,  1.0559,  ..., -1.0184, -0.0626, -0.3657]]],\n",
       "       grad_fn=<NativeLayerNormBackward0>), pooler_output=tensor([[-0.7910,  0.9982,  0.5162,  0.5645,  0.9931, -1.0000, -0.9679,  0.9552,\n",
       "         -0.9853,  0.9589, -0.9807,  0.9999,  0.5608, -0.9023,  0.3254,  0.9963,\n",
       "         -0.9993, -0.9995, -0.9986,  0.9893, -0.9805, -0.9412,  1.0000,  0.9999,\n",
       "         -0.9650,  0.9021, -0.9906,  0.9998,  0.9825, -0.9002, -0.0313,  0.9987,\n",
       "         -0.9999,  0.9993, -0.9979, -0.6498, -0.9986,  0.9922, -1.0000,  0.6821,\n",
       "          0.1280, -0.9983, -0.9985,  0.4523,  0.9548,  0.8372, -0.9949,  0.9992,\n",
       "         -0.8152, -0.9504,  0.9233,  0.9772,  0.9706, -0.9996,  0.9951,  0.9335,\n",
       "          0.9998,  0.4958, -0.9636, -0.9993, -0.4236,  0.9906, -0.8122,  0.8259,\n",
       "         -0.6304,  0.9495, -0.9998,  0.8438,  0.8735,  0.9972,  0.9967, -0.5223,\n",
       "          0.9734, -0.9767,  0.8727,  0.9219, -0.8907, -0.9994,  0.9975,  0.9954,\n",
       "          0.8217,  0.9813,  0.8811,  0.9990, -0.9982, -0.9557, -0.9945, -0.1240,\n",
       "          0.5805,  0.9986,  0.9990, -0.9980,  0.8054, -0.9962,  0.9970,  0.9105,\n",
       "          0.9781,  0.9947, -0.9689,  0.9968,  1.0000, -0.9918, -0.9983, -0.9989,\n",
       "          0.8950, -0.9991,  0.9992, -0.8593, -0.9647, -0.6932, -0.9995, -0.8731,\n",
       "          0.6549, -0.9997, -0.9844, -0.9018, -0.9893, -0.9721, -0.6195,  0.6840,\n",
       "          0.9653, -0.9998, -0.9830,  0.8567, -0.9991, -0.9945, -0.9942, -0.9987,\n",
       "         -0.8455,  0.9993,  0.9969, -0.9773,  0.9999,  0.9446, -0.9901,  0.9846,\n",
       "         -0.9995, -0.1405,  0.9344, -0.4566,  0.9992, -0.9883,  0.8935,  0.9972,\n",
       "          0.9795, -0.9471,  0.9444, -0.9386, -0.9981, -0.9994, -1.0000,  0.9651,\n",
       "          0.9739, -0.9875,  0.8699,  0.9993,  0.9994,  0.9557, -0.6019,  0.9648,\n",
       "          0.6508,  0.9806, -0.9633,  0.9558, -0.9883, -0.9997, -0.9367, -0.9662,\n",
       "         -0.9879, -0.9997,  0.9948, -0.9992, -0.9739, -0.9458, -0.9451,  0.9144,\n",
       "          0.9966,  0.9817,  0.9833,  0.9295,  1.0000, -0.6115,  0.9802,  0.4455,\n",
       "          0.7668, -0.9995,  0.7733, -0.8827, -0.9991, -0.8302, -0.8424,  0.9473,\n",
       "         -0.9886,  0.6797,  0.6204,  0.8177,  0.9167,  0.8128,  0.8469, -0.9969,\n",
       "          0.9989,  0.8881, -0.9991, -0.9994, -0.8558, -0.8194,  0.9279, -0.9474,\n",
       "         -0.9854,  0.9689, -0.9997, -0.9983, -0.9843,  0.2633, -0.8399, -0.8064,\n",
       "          0.9987,  0.9846,  0.9430, -0.9880,  1.0000, -0.9841, -0.9990, -0.8640,\n",
       "         -0.9979,  0.9163,  0.9979, -0.9998, -0.8630, -0.7422,  0.9035, -0.7474,\n",
       "          0.7429, -0.9756,  0.7962,  0.9832,  0.9982,  0.9991,  0.9469, -0.9969,\n",
       "          0.5806,  0.9687,  0.9999,  0.5612,  0.9999, -0.9670,  0.9994,  0.8751,\n",
       "         -0.6081, -0.9949,  0.9995, -0.9993, -0.8012,  0.9933,  0.9992, -0.9991,\n",
       "          0.9990,  0.9996,  0.9505, -0.9963,  0.9767, -0.5401,  0.7324,  0.7818,\n",
       "         -0.9856, -0.9573,  0.9754,  0.9958,  0.9892, -0.9942,  0.9960,  0.8151,\n",
       "         -0.9998, -0.9981, -0.9850,  0.9667, -0.9875,  0.8640, -0.9677,  0.9924,\n",
       "         -0.9588, -0.8726, -0.9618, -0.9460,  0.7362, -0.9723, -0.7751, -0.9454,\n",
       "          0.9993,  0.8828, -0.6301, -0.7683, -0.9798, -0.7620, -0.9996,  0.9755,\n",
       "          0.9709, -0.9947,  0.9898,  0.9919,  0.9973,  0.8937, -0.6852, -0.9992,\n",
       "          0.9452,  0.7104, -0.9023,  0.9998,  0.9022,  0.9968, -0.8535, -0.9982,\n",
       "         -0.9999,  0.9597,  0.9501, -0.9621, -0.9375,  0.3667,  0.7262, -0.9937,\n",
       "          0.9970, -0.9740, -0.7825,  0.9996, -0.6465, -0.9996, -0.9995, -0.6820,\n",
       "          0.9133,  0.6767, -0.3611,  0.9982,  0.9624,  0.9986,  0.7573,  0.9994,\n",
       "          0.9243,  0.9453,  0.9999, -0.9967, -0.8720, -0.9976,  0.9996,  0.9978,\n",
       "          0.9997,  0.9516,  0.9969, -0.9961,  0.7918,  0.9759, -0.7177,  0.7562,\n",
       "          0.8961,  0.9697,  0.9570,  0.9984, -0.9915,  0.9997,  0.9965,  0.8791,\n",
       "          0.8960, -0.9858,  0.6869,  0.8670, -0.8326, -0.9990,  0.8949, -0.8591,\n",
       "          0.9997, -0.9369, -0.9615, -0.8644, -0.6783,  0.9999, -0.9998,  0.9861,\n",
       "          0.9718,  0.9981, -0.9896,  0.9978,  0.9990,  0.9997,  0.9625, -0.9998,\n",
       "         -0.9644, -0.9995, -0.8790, -0.9993,  0.9963, -0.9971, -0.9992,  0.9450,\n",
       "         -0.7213, -0.9519, -0.9998,  0.9996,  0.8082,  0.8471,  0.9998, -0.9872,\n",
       "          0.8212, -0.9292, -0.9853,  0.9874, -0.9987, -0.3942, -0.9979, -0.8770,\n",
       "          0.6302, -0.9904,  0.9992,  0.9057,  0.9734,  0.9991,  0.9093, -0.9993,\n",
       "         -0.9853, -0.9851, -0.9800, -0.7828, -0.9993,  0.8048, -0.9489, -0.9989,\n",
       "         -0.9445,  0.9073, -0.9997,  0.9997, -0.6046, -0.8580, -0.7790,  0.8526,\n",
       "          0.9972,  0.9995, -0.8220,  0.9986,  0.9288,  0.9982, -0.0214, -0.7099,\n",
       "          0.8786, -0.2322, -0.9999, -0.9982,  0.9683,  0.9857, -0.8410, -0.6531,\n",
       "         -0.7315,  0.9989,  0.9975, -0.9677, -0.9369, -0.9962, -0.9969,  0.9706,\n",
       "          0.9830, -0.9098, -0.9932,  0.9575,  0.6775,  0.9566,  0.9996,  0.9994,\n",
       "         -0.9921,  0.9928, -0.9989,  0.9810, -0.7955,  0.9657,  0.8684, -0.9910,\n",
       "         -0.5694,  0.9335, -0.9993, -0.9999,  0.8650,  0.8981, -0.9674, -0.8482,\n",
       "          0.7646, -0.9866,  0.9979, -0.7495,  0.9998,  0.9998,  0.9979, -0.9663,\n",
       "         -0.5114, -0.9886, -0.9847, -0.9999, -0.9956, -0.9998, -0.7000, -0.9995,\n",
       "         -0.9943, -0.9985, -0.9986,  0.9678, -0.3102, -0.9984,  0.9991,  0.6377,\n",
       "          0.9963,  0.9989,  0.7082,  0.9867, -0.9419,  0.9893, -0.9636,  0.7840,\n",
       "          0.8296,  0.9993, -0.5786, -0.9415,  0.9752,  0.8513, -0.9986,  0.9998,\n",
       "          0.9984,  0.9988,  0.9999, -0.9900, -0.9948, -0.8826, -0.9707,  0.9999,\n",
       "         -0.7492,  0.7339,  0.9631, -0.8352, -0.9998,  0.8823, -0.9875,  0.9181,\n",
       "         -0.9998, -0.9941,  0.9472,  0.8855,  0.9996,  0.9998,  0.7425,  0.9115,\n",
       "         -0.9513,  0.9667, -0.9997,  0.9995,  0.9740,  0.9573,  1.0000, -0.9250,\n",
       "         -0.4410, -0.9981, -0.9494, -0.8091, -0.9938, -0.9972,  0.9996,  0.9999,\n",
       "          0.9195, -0.9988, -0.9992,  0.9996, -0.9958,  1.0000, -0.8097,  0.8539,\n",
       "         -0.8119,  0.8456, -0.7931, -0.9971,  0.9971, -0.8711, -0.9865, -0.9997,\n",
       "          0.9984, -0.9942, -0.9494,  0.8841, -0.9485, -0.9297,  0.9993, -0.7478,\n",
       "          0.5879,  0.9414,  0.9485, -0.9951, -0.6257,  0.7331,  1.0000, -0.9995,\n",
       "         -0.8834, -0.8563,  0.9556, -0.9997, -0.9948, -0.9774, -0.9501,  0.9999,\n",
       "         -0.7751,  0.9999, -0.6694,  0.9991,  0.9214,  0.9835,  0.9996,  0.9996,\n",
       "         -0.8783,  0.7090, -0.8799, -0.9941, -0.9947, -0.9938, -0.9999,  0.8887,\n",
       "         -0.9989, -0.8837, -0.9792, -0.9836,  0.9892,  0.8600,  0.9726,  0.9986,\n",
       "         -0.9999,  0.7985,  0.9987, -0.9842,  0.9705,  0.9052, -0.9991, -0.7734,\n",
       "          0.9999, -0.7774,  0.9421, -0.7789, -0.9988, -0.5484,  0.9676, -0.9986,\n",
       "          0.9973,  0.9967, -0.9239, -0.9874, -0.9927,  0.8134,  0.9793,  0.9284,\n",
       "          0.9995, -0.9450,  0.7950,  0.8434, -0.9838, -0.9545,  0.9991, -1.0000,\n",
       "         -0.7336,  0.9996,  0.9973, -0.9497,  0.7646, -0.9991, -0.9992,  0.9995,\n",
       "         -0.9999,  0.9888,  0.6886,  0.7962,  0.8895, -0.9995, -0.9735,  0.6907,\n",
       "         -0.8662, -0.9997,  1.0000,  0.9996, -0.9154,  0.9963,  0.9979, -0.9991,\n",
       "          0.9984,  0.9865,  0.9706,  0.9973,  0.9997, -0.9996, -0.9985, -0.9452,\n",
       "         -0.9990, -0.7147, -0.8982,  0.8155, -1.0000, -0.7998,  0.9930, -0.9853,\n",
       "          0.9990,  0.9970, -0.9977, -0.5622, -0.8914,  0.8672,  0.8218,  0.8915,\n",
       "         -0.9988,  0.8668,  0.9732,  0.9918, -0.9999,  0.5447,  0.6008, -0.8841,\n",
       "          0.9991, -0.9960,  0.4838, -0.9971,  0.9998, -0.9350, -0.8084, -0.9163,\n",
       "          0.9987,  0.9959, -0.9998,  0.9148,  0.9989,  0.9056,  0.9999,  0.7851,\n",
       "         -0.9261, -0.9973, -0.9552,  0.9115,  0.9999,  0.9973, -0.9973, -0.9999,\n",
       "          0.9990,  0.9996, -0.9996,  0.9991,  0.9814,  0.7837,  0.9992,  0.8452,\n",
       "          0.9989,  0.9061, -0.9903,  0.2583,  0.9987,  0.9764, -0.9919, -0.9506,\n",
       "         -0.7560, -0.9545, -0.8275,  0.9999, -0.9962, -0.9822,  0.9300,  0.9957,\n",
       "          0.9983, -0.9415,  0.9991,  0.8349, -0.9478,  0.9912,  0.9979, -0.9989]],\n",
       "       grad_fn=<TanhBackward0>), hidden_states=None, past_key_values=None, attentions=None, cross_attentions=None)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
